src/librustc_typeck/check/demand.rs

```diff
@@ -306,12 +304,181 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                                                &format!("{}, which {}", msg, depending_on_isize),
+                                                suggestion);
+                        }
+                        (_, None) => {
+                            err.span_suggestion(expr.span,
+                                                &format!("{}, which {}", msg, depending_on_usize),
+                                                suggestion);
+                        }
+                        _ => {
+                            err.span_suggestion(expr.span,
+                                                &format!("{}, which {}", msg, will_zero_extend),
+                                                suggestion);
+                        }
+                    }
+                    true
+                }
+                (&ty::TyUint(ref exp), &ty::TyInt(ref found)) => {
+                    match (found.bit_width(), exp.bit_width()) {
+                        (Some(found), Some(exp)) if found - 1 > exp => {
+                            err.span_suggestion(expr.span,
+                                                &format!("{}, which {}", msg, will_truncate),
+                                                suggestion);
+                        }
+                        (None, None) => {
+                            err.span_suggestion(expr.span,
+                                                &format!("{}, which {}", msg, will_sign_extend),
+                                                suggestion);
+                        }
+                        (None, _) => {
+                            err.span_suggestion(expr.span,
+                                                &format!("{}, which {}", msg, depending_on_usize),
+                                                suggestion);
+                        }
+                        (_, None) => {
+                            err.span_suggestion(expr.span,
+                                                &format!("{}, which {}", msg, depending_on_isize),
+                                                suggestion);
+                        }
+                        _ => {
+                            err.span_suggestion(expr.span,
+                                                &format!("{}, which {}", msg, will_sign_extend),
+                                                suggestion);
+                        }
+                    }
+                    true
+                }
+                (&ty::TyFloat(ref exp), &ty::TyFloat(ref found)) => {
+                    if found.bit_width() > exp.bit_width() {
+                        err.span_suggestion(expr.span,
+                                            &format!("{}, producing the closest possible value",
+                                                     msg),
+                                            suggestion);
+                        err.warn("casting here will cause Undefined Behavior if the value is \
```

Comment:
Should "undefined behaviour" be capitalised like this? [It doesn't seem to be](https://github.com/rust-lang/rust/blob/565907fefb2a69c279b59bffa84c6d30129cafe9/src/librustc_lint/builtin.rs#L1236) typical in error messages.